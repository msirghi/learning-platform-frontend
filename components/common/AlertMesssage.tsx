import React from 'react';
import Alert from '@material-ui/lab/Alert';
import { AlertType } from '../../common/enums';

/**
 * AlertMessage component props.
 */
type Props = {
  /**
   * Message that will be displayed
   */
  message: string;
  /**
   * Alert type. Depending on it component will have special colors.
   */
  type: AlertType;
};

/**
 * Component for displaying the alert messages. Uses Alert component from '@material-ui/lab' library.
 *
 * @version 0.1
 * @author [Sirghi Mihail](https://github.com/msirghi)
 */
export const AlertMessage: React.FC<Props> = ({ message, type }) => {
  return <Alert severity={type}>{message}</Alert>;
};
