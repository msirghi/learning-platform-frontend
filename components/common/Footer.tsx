import React from 'react';
import CssBaseline from '@material-ui/core/CssBaseline';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import Link from '@material-ui/core/Link';

function Copyright() {
  return (
    <Typography variant='body2'>
      {'Copyright Â© '}
      <Link color='inherit' href='https://material-ui.com/'>
        Your Website
      </Link>{' '}
      {new Date().getFullYear()}
      {'.'}
    </Typography>
  );
}

const useStyles = makeStyles((theme) => ({
  footer: {
    padding: theme.spacing(3, 2),
    marginTop: 'auto',
    backgroundColor: '#202020',
    color: '#fff'
  }
}));

export const SiteFooter = () => {
  const classes = useStyles();
  return (
    <footer className={classes.footer}>
      <Container maxWidth='sm'>
        <Typography variant='body1'>My sticky footer can be found here.</Typography>
        <Copyright />
      </Container>
    </footer>
  );
};
