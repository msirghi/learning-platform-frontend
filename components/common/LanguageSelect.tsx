import React, { useState } from 'react';
import LanguageIcon from '@material-ui/icons/Language';
import IconButton from '@material-ui/core/IconButton';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import Menu from '@material-ui/core/Menu';
import { useTranslation } from '../../i18n';
import { useDispatch } from 'react-redux';
import { setInterfaceLocale } from '../../redux/actions/preference/preferenceAction';

const languageChoice = [
  {
    key: 'en',
    value: 'English'
  },
  {
    key: 'ru',
    value: 'Русский'
  }
];

export const LanguageSelect: React.FC = () => {
  const [notificationMenuOpen, setNotificationMenuOpen] = useState<null | HTMLElement>(null);
  const { i18n } = useTranslation();
  const dispatch = useDispatch();

  const handleChange = (val: string) => {
    dispatch(setInterfaceLocale(val));
    i18n.changeLanguage(val);
    handleClose();
  };

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setNotificationMenuOpen(event.currentTarget);
  };

  const handleClose = () => {
    setNotificationMenuOpen(null);
  };

  return (
    <>
      <IconButton onClick={handleClick}>
        <LanguageIcon color='primary' />
      </IconButton>
      <Menu
        anchorEl={notificationMenuOpen}
        keepMounted
        open={Boolean(notificationMenuOpen)}
        onClose={handleClose}
      >
        <List>
          {languageChoice.map((item) => {
            return (
              <ListItem key={item.key} button onClick={() => handleChange(item.key)}>
                {item.value}
              </ListItem>
            );
          })}
        </List>
      </Menu>
    </>
  );
};
