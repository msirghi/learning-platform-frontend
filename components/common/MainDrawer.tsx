import React, { useEffect, useState } from "react";
import Drawer from "@material-ui/core/Drawer";
import IconButton from "@material-ui/core/IconButton";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemIcon from "@material-ui/core/ListItemIcon";
import ListItemText from "@material-ui/core/ListItemText";
import { makeStyles } from "@material-ui/core/styles";
import clsx from "clsx";
import { menuItems } from "./utils/menuItems";
import styles from "../../styles/modules/Shared.module.scss";
import { Hidden } from "@material-ui/core";

const drawerWidth = 240;

const useStyles = makeStyles((theme) => ({
  drawer: {
    width: drawerWidth,
  },
  drawerOpen: {
    background: theme.palette.primary.main,
    width: drawerWidth,
    transition: theme.transitions.create("width", {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen,
    }),
  },
  drawerClose: {
    background: theme.palette.primary.main,
    transition: theme.transitions.create("width", {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
    overflowX: "hidden",
    width: theme.spacing(8) + 1,
    [theme.breakpoints.up("sm")]: {
      width: 110,
    },
  },
  toolbar: {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    padding: theme.spacing(0, 1),
    ...theme.mixins.toolbar,
  },
  drawerPaper: {
    width: drawerWidth,
    background: theme.palette.primary.main,
  },
}));

type Props = {
  handleDrawerClose: () => void;
  open: boolean;
  mobileOpen: boolean;
};

export const MainDrawer: React.FC<Props> = ({ handleDrawerClose, open, mobileOpen }) => {
  const classes = useStyles();
  const [activeTab, setActiveTab] = useState<string | null>(null);

  useEffect(() => {
    /* istanbul ignore next */
    if (window) {
      setActiveTab(window.location.href);
    }
  }, []);

  if (!activeTab) {
    return <div />;
  }

  const drawerContent = () => {
    return (
      <>
        <div className={classes.toolbar}>
          <IconButton onClick={handleDrawerClose}>
            <img src='/images/register-logo-148.png' height={72} width={72} />
          </IconButton>
        </div>
        <List className={styles.mainMenuListContainer}>
          {menuItems.map((item) => {
            const isActiveTab = activeTab.includes(item.link);
            return (
              <div
                className={`${styles.mainMenuColor} ${isActiveTab && styles.mainMenuActive} ${
                  item.className
                }`}
                key={item.label}
              >
                <ListItem button>
                  <ListItemIcon className={styles.mainMenuIconContainer}>
                    <item.icon
                      className={`${isActiveTab ? styles.mainMenuActiveIcon : styles.mainMenuIcon}`}
                    />
                  </ListItemIcon>
                  {(open || mobileOpen) && (
                    <ListItemText
                      className={isActiveTab ? styles.mainMenuActiveText : ""}
                      primary={item.label}
                    />
                  )}
                </ListItem>
              </div>
            );
          })}
        </List>
      </>
    );
  };

  return (
    <>
      <Drawer
        variant='temporary'
        open={mobileOpen}
        onClose={handleDrawerClose}
        classes={{
          paper: classes.drawerPaper,
        }}
        ModalProps={{
          keepMounted: true,
        }}
      >
        {drawerContent()}
      </Drawer>
      <Hidden xsDown implementation='css'>
        <Drawer
          classes={{
            paper: clsx({
              [classes.drawerOpen]: open,
              [classes.drawerClose]: !open,
            }),
          }}
          className={clsx(classes.drawer, {
            [classes.drawerOpen]: open,
            [classes.drawerClose]: !open,
          })}
          variant='permanent'
          open={open}
        >
          {drawerContent()}
        </Drawer>
      </Hidden>
    </>
  );
};
