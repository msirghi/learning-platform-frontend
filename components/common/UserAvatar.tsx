import React from 'react';

type Props = {
  img?: string;
};

const styles = { height: 35, width: 35, background: '#bbb', borderRadius: '50%' };

export const UserAvatar: React.FC<Props> = ({ img }) => {
  if (img) {
    return <img style={styles} src={img} />;
  }

  return <span style={styles} />;
};
