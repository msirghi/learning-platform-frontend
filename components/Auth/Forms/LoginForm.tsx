import React, { FormEvent, useState } from 'react';
import { Button, LinearProgress, Typography, TextField } from '@material-ui/core';
import { useTranslation } from '../../../i18n';
import styles from '../../../styles/modules/Auth.module.scss';
import Link from '@material-ui/core/Link';
import { AlertType, AuthPage } from '../../../common/enums';

type Props = {
  onMessage: (type: AlertType, text: string) => void;
  onTabChange: (tab: AuthPage) => void;
};

function LoginForm({ onMessage, onTabChange }: Props) {
  const { t } = useTranslation();
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [isSubmitting, setSubmitting] = useState(false);

  const onSubmit = (e: FormEvent) => {
    e.preventDefault();
    setSubmitting(true);

    setTimeout(() => {
      setSubmitting(false);
      onMessage(AlertType.ERROR, t('auth:invalidCredentials'));
    }, 500);
  };

  return (
    <div>
      {isSubmitting && <LinearProgress />}
      <form onSubmit={onSubmit}>
        <TextField
          variant='outlined'
          name='email'
          label={t('auth:email')}
          inputProps={{
            'data-testid': 'email-field',
            'data-test': 'email',
            name: 'email'
          }}
          fullWidth
          margin='normal'
          value={email}
          onChange={(e) => setEmail(e.target.value)}
        />

        <TextField
          variant='outlined'
          name='password'
          label={t('auth:password')}
          inputProps={{
            'data-testid': 'password-field',
            'data-test': 'password',
            name: 'password'
          }}
          fullWidth
          type={'password'}
          margin='normal'
          value={password}
          onChange={(e) => setPassword(e.target.value)}
        />

        <Typography className={styles.link}>
          <Link href='#'>{t('auth:forgotPasswordLabel')}</Link>
        </Typography>

        <Typography className={styles.link}>
          {t('auth:dontHaveAnAccountLabel')}
          <Link onClick={() => onTabChange(AuthPage.REGISTER)} href='#'>
            {t('auth:signUp')}
          </Link>
        </Typography>

        <div className={styles.submitButton}>
          <Button
            type={'submit'}
            variant='contained'
            color='primary'
            data-testid='submit-button'
            data-test='submit-button'
            disabled={isSubmitting || !email || !password}
            className={styles.loginButton}
          >
            {t('auth:signIn')}
          </Button>
        </div>
      </form>
    </div>
  );
}

export default LoginForm;
