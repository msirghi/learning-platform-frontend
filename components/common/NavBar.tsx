import React, { useEffect, useState } from 'react';
import AppBar from '@material-ui/core/AppBar';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import Toolbar from '@material-ui/core/Toolbar';
import clsx from 'clsx';
import Typography from '@material-ui/core/Typography';
import NotificationsIcon from '@material-ui/icons/Notifications';
import styles from '../../styles/modules/Shared.module.scss';
import Divider from '@material-ui/core/Divider';
import { UserAvatar } from './UserAvatar';
import Badge from '@material-ui/core/Badge';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import { LanguageSelect } from './LanguageSelect';
import { useTranslation } from '../../i18n';

type Props = {
  classes: { [key: string]: string };
  handleDrawerOpen: () => void;
  handleDrawerClose: () => void;
  open: boolean;
};

export const NavBar: React.FC<Props> = ({ classes, handleDrawerOpen, open, handleDrawerClose }) => {
  const [activeTab, setActiveTab] = useState<string | undefined>();
  const { t, i18n } = useTranslation();

  useEffect(() => {
    if (global.window) {
      console.log('global.window.location :>> ', global.window.location.pathname);
      const href = global.window.location.pathname.replaceAll('/', '');
      setActiveTab(t(`common:headerTabs.${href}`));
    }
  }, [global.window, i18n]);

  return (
    <AppBar
      position='fixed'
      className={clsx(classes.appBar, {
        [classes.appBarShift]: open
      })}
    >
      <Toolbar>
        <IconButton
          color='inherit'
          aria-label='open drawer'
          onClick={handleDrawerOpen}
          edge='start'
          className={clsx(classes.menuButton, {
            [classes.hide]: open
          })}
        >
          <MenuIcon />
        </IconButton>

        <IconButton
          color='inherit'
          aria-label='close drawer'
          onClick={handleDrawerClose}
          edge='start'
          className={clsx(classes.menuButton, {
            [classes.hide]: !open
          })}
        >
          <ChevronLeftIcon />
        </IconButton>
        <Typography variant='h6' noWrap>
          {activeTab}
        </Typography>
        <div className={styles.navBarRightIconsContainer}>
          <LanguageSelect />
          <IconButton>
            <Badge badgeContent={4} color='primary'>
              <NotificationsIcon color={'primary'} />
            </Badge>
          </IconButton>
          <IconButton>
            <UserAvatar />
          </IconButton>
        </div>
      </Toolbar>
      <div className={styles.navBarDividerContainer}>
        <Divider light />
      </div>
    </AppBar>
  );
};
