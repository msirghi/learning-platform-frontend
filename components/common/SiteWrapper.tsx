import React, { useEffect } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import { CssBaseline } from '@material-ui/core';
import { NavBar } from './NavBar';
import { MainDrawer } from './MainDrawer';
import { SiteFooter } from './Footer';
import { useWindowSize } from './hooks/useWindowResize';

const drawerWidth = 240;

const useStyles = makeStyles((theme) => ({
  wrapper: {
    display: 'flex',
    flexDirection: 'column',
    minHeight: '100vh'
  },
  root: {
    display: 'flex'
  },
  main: {
    marginTop: theme.spacing(8),
    marginBottom: theme.spacing(2)
  },
  appBar: {
    backgroundColor: '#fafafa',
    color: '#000',
    boxShadow: 'none',
    [theme.breakpoints.up('sm')]: {
      width: `calc(100% - ${drawerWidth / 2}px)`
    }
  },
  appBarShift: {
    width: `calc(100% - ${drawerWidth}px)`,
    transition: theme.transitions.create(['width', 'margin'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen
    })
  },
  menuButton: {
    marginRight: 36
  },
  hide: {
    display: 'none'
  },
  toolbar: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'flex-end',
    padding: theme.spacing(0, 1),
    ...theme.mixins.toolbar
  },
  content: {
    flexGrow: 1,
    padding: theme.spacing(3)
  }
}));

export const SiteWrapper: React.FC = ({ children }) => {
  const classes = useStyles();
  const [desktopOpen, setDesktopOpen] = React.useState(false);
  const [mobileOpen, setMobileOpen] = React.useState(false);
  const [width] = useWindowSize();

  useEffect(() => {
    if (width < 768 && desktopOpen) {
      setDesktopOpen(false);
      return;
    }

    if(width > 768 && mobileOpen) {
      setMobileOpen(false);
    }
  }, [width]);

  const handleDrawerOpen = () => {
    if (width > 768) {
      setDesktopOpen(true);
      return;
    }
    setMobileOpen(true);
  };

  const handleDrawerClose = () => {
    if (width > 768) {
      setDesktopOpen(false);
      return;
    }
    setMobileOpen(false);
  };

  return (
    <div className={classes.wrapper}>
      <div className={classes.root}>
        <CssBaseline />
        <NavBar
          classes={classes}
          handleDrawerClose={handleDrawerClose}
          handleDrawerOpen={handleDrawerOpen}
          open={desktopOpen}
        />
        <MainDrawer
          handleDrawerOpen={handleDrawerOpen}
          mobileOpen={mobileOpen}
          handleDrawerClose={handleDrawerClose}
          open={desktopOpen}
        />
        <main className={classes.content}>
          <div className={classes.toolbar} />
          {children}
        </main>
      </div>
      <SiteFooter />
    </div>
  );
};
