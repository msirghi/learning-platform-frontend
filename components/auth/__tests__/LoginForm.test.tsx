import { Button, LinearProgress } from '@material-ui/core';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { findByTestAttr } from '../../../common/testUtils';
import LoginForm from '../forms/LoginForm';

describe('LoginForm component', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<LoginForm onMessage={jest.fn()} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should have disabled button by default', () => {
    const wrapper = shallow(<LoginForm onMessage={jest.fn()} />);
    const button = wrapper.find(Button);

    expect(button).toHaveLength(1);
    expect(button.props().disabled).toBeTruthy();
  });

  it('should have enabled button on fullfiled fields', () => {
    const { getByTestId } = render(<LoginForm onMessage={jest.fn()} />);

    const emailField = getByTestId('email-field');
    expect(emailField).toBeInTheDocument();

    const passwordField = getByTestId('password-field');
    expect(passwordField).toBeInTheDocument();

    const button = getByTestId('submit-button');
    expect(button).toBeDisabled();

    fireEvent.change(emailField, { target: { value: 'email@email.com' } });
    fireEvent.change(passwordField, { target: { value: 'password' } });
    expect(button).toBeEnabled();
    fireEvent.click(button);
  });

  it('should submit the form and show LinearProgress', () => {
    const wrapper = mount(<LoginForm onMessage={jest.fn()} />);

    const emailField = findByTestAttr(wrapper, 'email');
    expect(emailField).toHaveLength(1);

    const passwordField = findByTestAttr(wrapper, 'password');
    expect(passwordField).toHaveLength(1);

    emailField.simulate('change', 'email@email.com');
    passwordField.simulate('change', 'password');

    const form = wrapper.find('form');
    expect(form).toHaveLength(1);

    form.simulate('submit');
    expect(wrapper.find(LinearProgress)).toHaveLength(1);
  });
});
