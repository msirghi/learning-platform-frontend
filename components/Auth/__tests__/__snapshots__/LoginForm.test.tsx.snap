// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component should match snapshot 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      inputProps={
        Object {
          "data-test": "email",
          "data-testid": "email-field",
          "name": "email",
        }
      }
      label="auth:email"
      margin="normal"
      name="email"
      onChange={[Function]}
      value=""
      variant="outlined"
    />
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      inputProps={
        Object {
          "data-test": "password",
          "data-testid": "password-field",
          "name": "password",
        }
      }
      label="auth:password"
      margin="normal"
      name="password"
      onChange={[Function]}
      type="password"
      value=""
      variant="outlined"
    />
    <WithStyles(ForwardRef(Typography))
      className="link"
    >
      <WithStyles(ForwardRef(Link))
        href="#"
      >
        auth:forgotPasswordLabel
      </WithStyles(ForwardRef(Link))>
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      className="link"
    >
      auth:dontHaveAnAccountLabel
      <WithStyles(ForwardRef(Link))
        href="#"
      >
        auth:signUp
      </WithStyles(ForwardRef(Link))>
    </WithStyles(ForwardRef(Typography))>
    <div
      className="submitButton"
    >
      <WithStyles(ForwardRef(Button))
        className="loginButton"
        color="primary"
        data-test="submit-button"
        data-testid="submit-button"
        disabled={true}
        type="submit"
        variant="contained"
      >
        auth:signIn
      </WithStyles(ForwardRef(Button))>
    </div>
  </form>
</div>
`;
